src/run.rs

```diff
@@ -0,0 +1,202 @@
+
+        let mut cmd_handle = cmd_handle.unwrap();
+
+        let std_out = cmd_handle.stdout.take();
+        let std_err = cmd_handle.stderr.take();
+        let mut std_out_handle = None;
+        let mut std_err_handle = None;
+
+        let shared_handle = Arc::new(Mutex::new(cmd_handle));
+
+        let child_clone = shared_handle.clone();
+        let kill_trigger_clone = kill_trigger.clone();
+        thread::spawn(move || kill_thread(&kill_trigger_clone, child_clone));
+
+        if let Some(output) = std_out {
+            let mut buffered_stdout = BufReader::new(output);
+            let new_name = command_name.clone();
+            let new_chan = send_chan.clone();
+            std_out_handle = Some(thread::spawn(move || {
+                read_stream(&new_name, new_chan, &mut buffered_stdout, true);
+            }));
+        }
+
+        if let Some(output) = std_err {
+            let mut buffered_stdout = BufReader::new(output);
+            let new_name = command_name.clone();
+            let new_chan = send_chan.clone();
+            std_err_handle = Some(thread::spawn(move || {
+                read_stream(&new_name, new_chan, &mut buffered_stdout, false);
+            }));
+        }
+
+        if let Some(handle) = std_out_handle {
+            let _ = handle.join();
+        }
+
+        if let Some(handle) = std_err_handle {
+            let _ = handle.join();
+        }
+
+        let exit_status = shared_handle.lock().unwrap().wait();
+        match exit_status {
+            Ok(status) => {
+                let _ = send_chan.send(OutputMessage {
+                    name: command_name.clone(),
+                    message: OutputMessagePayload::Done(status.code()),
+                });
+
+                match options.restart {
+                    RestartOptions::Continue => {
+                        return (command_name, Some(status));
+                    }
+                    RestartOptions::Restart => {
+                        if status.success() {
+                            return (command_name, Some(status));
+                        }
+                    }
+                    RestartOptions::Kill => {
+                        if !status.success() {
+                            let _ = kill_trigger.initiate_kill();
+                        }
+                        return (command_name, Some(status));
+                    }
+                };
+            }
+            Err(e) => {
+                let _ = send_chan.send(OutputMessage {
+                    name: command_name.clone(),
+                    message: OutputMessagePayload::Error(e),
+                });
+                return (command_name, None);
+            }
+        }
+    })
+}
+
+fn kill_thread(kill_trigger: &kill_barrier::KillBarrier, child: Arc<Mutex<process::Child>>) {
+    let _ = kill_trigger.wait();
+
+    let lock_res = child.lock();
+    if let Ok(mut locked_child) = lock_res {
+        let _ = locked_child.kill();
+    }
```

Comment:
What happens if it cannot acquire a lock on `child`? Does it not kill it then?

Reply:
I failed lock acquisition comes from a poisoned lock. In rust land, that means that some other thread has panicked while holding that lock. 
In that case, I let that panic take charge so as to not pollute the stack trace

Reply:
Ah, I forgot about the poisoned lock issue. Thanks for the explanation!